modules: introducing MODULE_ROOT to allow glue code in Zephyr repo (#30904)
zephyrproject-rtos/zephyr: doc/guides/modules.rst

     cmake: .
     kconfig: Kconfig

.. _modules_build_files_extern:

> Build files in a module root

Author: Well, this is the title for the section.

You cannot expect a term to be know in the title, as the title reflects what will be describe.

Maybe just have `Module root` as the title ?
```suggestion
Module root
```

> Build files in a module root

Author: Maybe move the section down ?

============================

Inclusion of build files located outside of the Zephyr module itself, can be
described as:

.. code-block:: yaml

   build:
     cmake-extern: True
     kconfig-extern: True

This allows control of the build inclusion to be described externally to the
Zephyr module.

> The Zephyr repository itself is always added as a Zephyr module root.

Reviewer: I suggest pointing to the section about module roots here.

Author: Good point, will do.

Author: section re-wriiten and moved.

Link added in beginning of the section.

The Zephyr repository contain :file:`CMakeLists.txt` and :file:`Kconfig` build
files for certain known Zephyr modules.

Those build files are located under

.. code-block:: none

   <ZEPHYR_BASE>
   └── modules
       └── <module_name>
           ├── CMakeLists.txt
           └── Kconfig.<module_name>

You can create a similar ``MODULE_ROOT`` for additional modules, and make those
modules known to Zephyr.

Create a ``MODULE_ROOT`` with the following structure

.. code-block:: none

   <MODULE_ROOT>
   └── modules
       ├── modules.cmake
       └── <module_name>
           ├── CMakeLists.txt
           └── Kconfig.<module_name>

and then build your application by specifying ``-DMODULE_ROOT`` parameter to the
CMake build system. The ``MODULE_ROOT`` accepts a CMake list of roots as
argument.

.. note::

   ``ZEPHYR_BASE`` is always added as a ``MODULE_ROOT`` with the lowest priority.
   This allows you to overrule any build files under
   ``<ZEPHYR_BASE>/modules/<module_name>`` with your own implementation your own
   ``MODULE_ROOT``.

The :file:`modules.cmake` file must contain the logic that specifies the build
files for Zephyr modules via specifically named CMake variables.

> To include a module's CMake file, the variable ``ZEPHYR_<MODULE_NAME>_CMAKE_DIR``
> is used.

Reviewer: ```suggestion
To include a module's CMake file, set the variable ``ZEPHYR_<MODULE_NAME>_CMAKE_DIR``
to the path containingh the CMake file.
```

Author: Thanks, applied.

> To include a module's Kconfig file, the variable ``ZEPHYR_<MODULE_NAME>_KCONFIG``
> is used.

Reviewer: ```suggestion
To include a module's Kconfig file, set the variable ``ZEPHYR_<MODULE_NAME>_KCONFIG``
to the path to the Kconfig file.
```

Author: Thanks, applied.

The following is an example on how to add support the the ``FOO`` module.

Create the following structure

.. code-block:: none

   <MODULE_ROOT>
   └── modules
       ├── modules.cmake
       └── foo
           ├── CMakeLists.txt
           └── Kconfig.foo

and inside the :file:`modules.cmake` file, add the following content

.. code-block:: cmake

   set(ZEPHYR_FOO_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR}/foo)
   set(ZEPHYR_FOO_KCONFIG   ${CMAKE_CURRENT_LIST_DIR}/foo/Kconfig.foo)

.. _modules_build_files_extern:

> Build files in a module root

Reviewer: The title reads a bit wonky since it's unclear whether "build" is a verb or adjective. I'm not sure this even needs to be a separate section since it's so related to the previous.

Author: >  I'm not sure this even needs to be a separate section since it's so related to the previous.

If you look at the menu, it should have it's own section:
![image](https://user-images.githubusercontent.com/9802583/103763028-8888c000-5019-11eb-9ec5-34a884f59a3e.png)

That will make it visible that there are two sections:
- `Build files` 
- `Build files in module root`
So depending on what you need to do, you can go to the right section, otherwise a user might just go to the `Build files` section  when they actually needed the `Build files in module root`.

Reviewer: Ok
```suggestion
Build files placed in a module root
```

Author: Choosing the first version, as `Build files` is the important part of the info the reader need to notice.

  include(${ZEPHYR_CURRENT_MODULE_DIR}/cmake/code.cmake)

.. _modules_module_root:

> Module root

Reviewer 2: need to change that from "Module root" to whatever module_ext_root means now.

Author: Done, including an update to `Build files placed in a module root`below.

> Module ext root

Reviewer 2: what does ext mean?. Can you spell this out? :)

Author: Please take a look at updated docs.

  include(${ZEPHYR_CURRENT_MODULE_DIR}/cmake/code.cmake)

.. _modules_module_root:

Module root
===========

> Zephyr repository contains build files certain known Zephyr modules.

Author: Thanks.

> You can add additional paths with build files for Zephyr module by using the

Author: Good point, will try to rephrase.

> You can add additional paths with build files for Zephyr module by using the

Author: Rephrased.

``-DMODULE_ROOT`` parameter to the CMake build system. This will tell the build

> system to also search this particular path for Zephyr modules using

Reviewer: Can you clarify whether it is possible to add multiple additional module roots?

Author: It is.
Will highlight that when updating the description.

Author: done

``cmake-extern: True`` or ``kconfig-extern: True`` settings as described in
:ref:`modules_build_files_extern`.

Each additional ``MODULE_ROOT`` must contain a :file:`modules.cmake` file which

> must contain the logic that specifies the build files for Zephyr modules.

Author: Thanks.

The file must be located as:

.. code-block:: none

   <MODULE_ROOT>
   └── modules

>        └── modules.cmake

Author: Right, will rephrase the text, cause the `MODULE_ROOT` should have same behavior as `BOARD / SOC / ARCH_ROOT`s.

>        └── modules.cmake

Author: done

In the Zephyr repository the build files for Zephyr modules are given names
according to the following pattern:

.. code-block:: none

   <ZEPHYR_BASE>
   └── modules
       └── <module_name>
           ├── CMakeLists.txt
           └── Kconfig.<module_name>

For user defined ``MODULE_ROOT`` it is advised to follow a similar structure,
but you may decide to use a custom pattern.
A custom pattern can be implemented in the

> :file:`<MODULE_ROOT>/modules/modules.cmake` file.

Reviewer: >Each additional ``MODULE_ROOT`` must contain a :file:`modules.cmake`

IMO implies

```suggestion
:file:`<MODULE_ROOT>/modules.cmake` file.
```

Author: Section description updated to make this clearer.

- ``arch_root``: Contains additional architectures that are available to the
  build system. Additional architectures must be located in a
  :file:`<arch_root>/arch` folder.

> - ``module_root``: Contains :file:`CMakeLists.txt` and :file:`Kconfig` files for

Reviewer: I think this needs a bit more explanation, perhaps in the Module root section.

Author: Not here, as this simply describes what additional roots that can be defined in the module.yml file.

Who those roots are used is described elsewhere, as you also suggest, in the Module root section.
Not sure what you want to describe though. Content of the CMakeLists.txt file and Kconfig file ?

There is no difference between a CMakeLists.txt or Kconfig file here or elsewhere in the build system.
Only difference, is that they are only included into the build system when the module is available.

So would be outside the scope of modules to describe the CMake or Kconfig content.

Reviewer: I can't see that it's mentioned in the module root section. It just needs a sentence e.g. "Additional module roots can also be added in the module.yml file, see <ref>."

Author: ahh, now I get it :smile: .
Will do, thanks.

Author: done
